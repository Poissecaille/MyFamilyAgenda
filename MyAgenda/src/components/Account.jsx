import React from 'react';
import { colors } from '../utility/Global';
import {
  StyleSheet, TouchableOpacity, Text, Image, View
} from 'react-native';

const styles = StyleSheet.create({
  backgroundAppImage: {
    width: '100%',
    height: '100%',
    position: 'absolute',
  },
  mainView: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    color: colors.White,
    fontSize: 40,
    fontFamily: 'PlayfairDisplay-SemiBoldItalic',
    paddingBottom: 30,
    textShadowColor: 'black',
    textShadowOffset: { width: 0, height: 0 },
    textShadowRadius: 20,
  },
  authButton: {
    padding: 15,
    backgroundColor: colors.Clearer3,
    borderRadius: 5,
    marginBottom: 15
  },
  textButton: {
    fontFamily: 'Roboto_700Bold',
    textAlign: 'center',
    color: colors.Black
  },
  menuOpacity: {
    backgroundColor: 'rgba(255, 255, 255, 0.3)',
    padding: 40,
    marginTop: 10
  }
});

export const AccountScreen = (props) => {
  return <>
    <Image style={styles.backgroundAppImage} source={require('../assets/family.jpg')} />
    <View style={styles.mainView}>
      <Text style={styles.title}>My Family Calendar</Text>
      <View style={styles.menuOpacity}>
        <TouchableOpacity style={styles.authButton} onPress={() => props.navigation.navigate('Login')}>
          <Text style={styles.textButton}>LOGIN</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.authButton} onPress={() => props.navigation.navigate('Register')}>
          <Text style={styles.textButton}>REGISTER</Text>
        </TouchableOpacity>
      </View>
    </View>
  </>
}
