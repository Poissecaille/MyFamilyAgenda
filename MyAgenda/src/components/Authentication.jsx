import React, { useState, createContext } from 'react';
import firebase from 'firebase/compat/app';
import 'firebase/compat/auth';
import 'firebase/compat/firestore';
import { getDatabase, ref, set, push } from 'firebase/database';

export const AuthenticationContext = createContext();
const db = getDatabase();

const loginRequest = (email, password) =>
  firebase.auth().signInWithEmailAndPassword(email, password);


export const AuthenticationContextProvider = ({ children }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [user, setUser] = useState(null);
  const [error, setError] = useState(null);

  function writeUserData(email) {
    const reference = ref(db, 'Users/');
    const newReference = push(reference) // generate new id under Users/
    set(newReference, {
      email: email,
    });
  }

  function createBaseAgendas(email) {
    const reference = ref(db, 'Agendas/');
    let newReference = push(reference); //generate new id under Agendas/
    set(newReference, {
      name: "Mon agenda familial",
      owner: email,
      members: {},
      events: {},
    });
    newReference = push(reference); //generate new id under Agendas/
    set(newReference, {
      name: "Mon agenda personnel",
      owner: email,
      members: {},
      events: {},
    });
  }



  firebase.auth().onAuthStateChanged(usr => {
    if (usr) {
      setUser(usr);
      setIsLoading(false);
    } else {
      setIsLoading(false);
    }
  });

  const onLogin = (email, password) => {
    setIsLoading(true);
    loginRequest(email, password)
      .then(u => {
        setUser(u);
        setIsLoading(false);
      })
      .catch(e => {
        setIsLoading(false);
        setError(e.toString());
      });
  };

  const onRegister = (email, password, repeatedPassword) => {
    setIsLoading(true);
    if (password !== repeatedPassword) {
      setError('Error: Passwords do not match');
      return;
    }
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(u => {
        writeUserData(email);
        createBaseAgendas(email)
        setUser(u);
        setIsLoading(false);
      })
      .catch(e => {
        setIsLoading(false);
        setError(e.toString());
      });
  };

  const onLogout = () => {
    firebase
      .auth()
      .signOut()
      .then(() => {
        setUser(null);
        setError(null);
      });
  };

  return (
    <AuthenticationContext.Provider
      value={{
        isAuthenticated: !!user,
        user,
        isLoading,
        error,
        onLogin,
        onRegister,
        onLogout,
      }}>
      {children}
    </AuthenticationContext.Provider>
  );
};
