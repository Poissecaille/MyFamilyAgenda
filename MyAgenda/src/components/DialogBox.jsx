import { View, StyleSheet, Text, Image, TouchableOpacity, Animated } from "react-native"
import React, { useState, useEffect, useRef } from 'react';
import { colors, discordSounds, random } from "../utility/Global";
let Sound = require('react-native-sound');


export const DialogBox = (props) => {
    const [text, setText] = useState(JSON.stringify(props.textArray));
    const [index, setIndex] = useState(0);
    const [restartAnimation, setRestartAnimation] = useState(false);
    const startValue = new Animated.Value(-5)
    const endValue = 5

    useEffect(() => {
        Animated.loop(
            Animated.spring(startValue, {
                toValue: endValue,
                bounciness: 2,
                speed: 2,
                useNativeDriver: true,
            }),
            { iterations: 5 },
        ).start();
    }, [])

    useEffect(() => {
        let whoosh = new Sound(discordSounds[random(discordSounds)], Sound.MAIN_BUNDLE, (error) => {
            if (error) {
                console.log('failed to load the sound', error);
            } else {
                whoosh.play((success) => {
                    if (success) {
                        whoosh.release();
                    }
                })
            }
        });
        Animated.loop(
            Animated.spring(startValue, {
                toValue: endValue,
                bounciness: 2,
                speed: 2,
                useNativeDriver: true,
            }),
            { iterations: 5 },
        ).start();
    }, [restartAnimation])

    const handleOnClick = () => {
        setIndex(index + 1);
        setRestartAnimation(!restartAnimation);
        props.resetDiscord();
        if (index >= props.textArray.length - 1) {
            props.resetDialog();
            setIndex(0);
        }
    };

    const styles = StyleSheet.create({
        bull: {
            backgroundColor: colors.White,
            padding: 20,
            borderRadius: 60,
            borderColor: colors.Black,
            borderWidth: 2,
            flexDirection: 'row',
        },
        innerBull1: {
            flex: 10
        },
        innerBull2: {
            flex: 1,
            transform: [
                { translateY: startValue }
            ]
        },
        textBull: {
            fontFamily: 'Roboto_500Medium',
            textAlign: 'center',
            color: colors.Black
        }
    });
    return (
        <>
            <TouchableOpacity style={styles.bull}
                onPress={handleOnClick}
            >
                <View style={styles.innerBull1}>
                    <Text style={styles.textBull}>{JSON.parse(text)[index]}</Text>
                </View>
                <Animated.View style={styles.innerBull2}>
                    <Image source={require('../assets/ic_touch_app.png')} />
                </Animated.View>
            </TouchableOpacity>
        </>
    )
}