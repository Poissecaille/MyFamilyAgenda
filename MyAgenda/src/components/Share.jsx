import React from 'react';
import { Share, Image, TouchableOpacity, StyleSheet, View } from 'react-native';
import { colors } from '../utility/Global';

export const ShareComponent = () => {

  const styles = StyleSheet.create({
    shareView: {
      position: 'absolute',
      width: '100%',
      height: '100%',
      flexDirection: 'column',
      justifyContent: 'flex-end',
    },
    shareTouchable: {
      justifyContent: 'center',
      textAlign: 'center',
      backgroundColor: colors.Clearer3,
      height: 60,
      width: 60,
      borderRadius: 20,
      justifyContent: 'center',
      alignSelf: 'flex-end'
    },
    shareImage: {
      aspectRatio: 1,
      width: '50%',
      height: '50%',
      alignSelf: 'center'
    }
  })

  const onShare = async () => {
    try {
      const result = await Share.share({
        title: 'App link',
        message: 'Please install this app and stay safe , AppLink :https://play.google.com/store/apps/details?id=nic.goi.aarogyasetu&hl=en',
        url: 'https://play.google.com/store/apps/details?id=nic.goi.aarogyasetu&hl=en'
      });
      if (result.action === Share.sharedAction) {
        if (result.activityType) {
          // shared with activity type of result.activityType
        } else {
          // shared
        }
      } else if (result.action === Share.dismissedAction) {
        // dismissed
      }
    } catch (error) {
      alert(error.message);
    }
  };
  return (
    <View style={styles.shareView}>
      <TouchableOpacity onPress={onShare} style={styles.shareTouchable}>
        <Image
          style={styles.shareImage}
          source={require('../assets/share2.png')}
        />
      </TouchableOpacity>
    </View>
  );
};