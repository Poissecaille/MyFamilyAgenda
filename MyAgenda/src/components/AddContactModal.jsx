import React, { useState, useContext } from 'react';
import { AuthenticationContext } from '../components/Authentication';
import { getDatabase, ref, get, push, set } from 'firebase/database';
import { StyleSheet, View, Modal, Text, TouchableOpacity, Button, TextInput } from 'react-native';
import { colors } from '../utility/Global';
import { screenSize } from '../utility/Global';
import { RadioButton } from 'react-native-paper';


const styles = StyleSheet.create({
  placeholder: {
    color: colors.White
  },
  arrowIcon: {
    color: colors.White
  },
  addAgenda: {
    backgroundColor: colors.Primary_Color,
    alignItems: 'center',
    width: '30%',
    padding: 10,
    borderRadius: 10,
    left: 250,
    bottom: 45
  },
  addAgendaText: {
    color: colors.White,
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 22,
  },
  modalView: {
    margin: 20,
    backgroundColor: 'white',
    borderRadius: 20,
    padding: 50,
    width: (screenSize.width * 3) / 4,
    height: (screenSize.height * 2) / 4,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  exitButton: {
    position: 'absolute',
    top: 10,
    right: 10,
    borderRadius: 20,
  },
  create: {
    backgroundColor: colors.Darker1,
    padding: 20,
    borderRadius: 20
  },
  createText: {
    color: colors.White
  }

});

export const AddContactModal = (props) => {
  const [contactEmail, setContactEmail] = useState("")
  const [type, setType] = React.useState('friendRequest');
  const { user } = useContext(AuthenticationContext);


  async function checkUser() {
    const db = getDatabase();
    const userRef = ref(db, 'Users/');
    const res = get((userRef)).then(snap => {
      const users = snap.val();
      const keys = Object.keys(users);
      for (const key of keys) {
        const user = users[key];
        if (user.email === contactEmail) {
          return user.email;
        }
      }
      return null
    }).catch(error => {
      console.log(error);
    })
    return res;
  }

  const sendContactRequest = async () => {
    console.log("sendContactRequest...");
    const targetEmail = await checkUser(); // find user
    if (!targetEmail) {
      alert("Cet utilisateur n'est pas enregistré. Pensez à l'inviter à rejoindre l'application ! :)");
      props.onRequestClose();
      return
    }
    if (targetEmail == user.email) {
      alert("Vous ne pouvez pas vous envoyer de demande d'ami à vous-même !");
      props.onRequestClose();
      return
    }
    if (targetEmail) {
      console.log("found user, sending request");
      // Créer la notification
      const db = getDatabase();
      let notifRef = ref(db, 'Notifications/');
      const newRef = push(notifRef)
      set(newRef, {
        'message': type === "friendRequest" ? `${user.email} veut vous ajouter en tant qu'ami(e)` :
          `${user.email} veut vous ajouter en tant que membre de sa famille`,
        'target': targetEmail,
        'origin': user.email,
        'type': type
      })
      alert("Demande envoyée !");
    }
    props.onRequestClose()
    return
  }


  return (
    <>

      <Modal
        animationType="fade"
        transparent={true}
        visible={props.visible}
        onRequestClose={() => {
          alert('Modal has been closed.');
          props.onRequestClose();
        }}
      >
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <View style={styles.exitButton}>
              <Button
                title="X"
                color={colors.Darker1}
                onPress={() => props.onRequestClose()}
              />
            </View>
            <TextInput
              onChangeText={setContactEmail}
              value={contactEmail}
              placeholder="Contact email"
              keyboardType="email-address"
            />
            <Text> Ami </Text>
            <RadioButton
              value="friendRequest"
              status={type === 'friendRequest' ? 'checked' : 'unchecked'}
              onPress={() => setType('friendRequest')}
            />
            <Text> Famille </Text>
            <RadioButton
              value="familyRequest"
              status={type === 'familyRequest' ? 'checked' : 'unchecked'}
              onPress={() => setType('familyRequest')}
            />

            <TouchableOpacity style={styles.create} onPress={sendContactRequest}>
              <Text style={styles.createText}> Add contact </Text>
            </TouchableOpacity>
          </View>
        </View>
      </Modal>
    </>
  );
}