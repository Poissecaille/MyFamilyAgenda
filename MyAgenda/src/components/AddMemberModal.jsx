import React, { useState, useEffect, useContext } from 'react';
import { AuthenticationContext } from '../components/Authentication';
import { getDatabase, ref, get, push, set } from 'firebase/database';
import { StyleSheet, View, Modal, Text, TouchableOpacity, Button } from 'react-native';
import { colors } from '../utility/Global';
import { screenSize } from '../utility/Global';



const styles = StyleSheet.create({
  placeholder: {
    color: colors.White
  },
  arrowIcon: {
    color: colors.White
  },
  addAgenda: {
    backgroundColor: colors.Primary_Color,
    alignItems: 'center',
    width: '30%',
    padding: 10,
    borderRadius: 10,
    left: 250,
    bottom: 45
  },
  addAgendaText: {
    color: colors.White,
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 22,
  },
  modalView: {
    margin: 20,
    backgroundColor: 'white',
    borderRadius: 20,
    padding: 50,
    width: (screenSize.width * 3) / 4,
    height: (screenSize.height * 2) / 4,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  exitButton: {
    position: 'absolute',
    top: 10,
    right: 10,
    borderRadius: 20,
  },
  create: {
    backgroundColor: colors.Darker1,
    padding: 20,
    borderRadius: 20
  },
  createText: {
    color: colors.White
  },
  relationBox: {
    borderWidth: 2,
    padding: 20,
    borderBottomLeftRadius: 20,
    borderBottomRightRadius: 20
  }

});

export const AddMemberModal = (props) => {

  //const [contactEmail,setContactEmail] = useState("")
  const { user } = useContext(AuthenticationContext);
  const [relations, setRelations] = useState([])

  async function checkUser(email) {
    const db = getDatabase();
    const userRef = ref(db, 'Users/');
    const res = get((userRef)).then(snap => {
      const users = snap.val();
      const keys = Object.keys(users);
      for (const key of keys) {
        const user = users[key];
        if (user.email === email) {
          return user.email;
        }
      }
      return null
    }).catch(error => {
      console.log(error);
    })
    return res;
  }

  useEffect(() => {
    getRelationList();
  })
  const getReferenceByEmail = async (email) => {
    const db = getDatabase();
    const userRef = ref(db, 'Users/');
    return get((userRef)).then(snap => {
      const users = snap.val();
      const keys = Object.keys(users);
      for (const key of keys) {
        const user = users[key];
        if (user.email === email) {
          return key;
        }
      }
    }).catch(error => {
      console.log(error);
    })
  }

  const getRelationList = async () => {
    const db = getDatabase();

    const myId = await getReferenceByEmail(user.email);
    const relationRef = ref(db, `Users/${myId}/relations/`);
    const res = await get((relationRef)).then(snap => {
      const relations = snap.val();
      return Object.keys(relations);
    }).catch(error => {
      console.log(error);
    })
    const userRef = ref(db, 'Users/');
    const resUsers = get((userRef)).then(snap => {
      const users = snap.val();
      const list = [];
      for (const key of res) {
        list.push(users[key].email);
      }
      return list;
    }).catch(error => {
      console.log(error);
    })



    setRelations(await resUsers);
  }

  const sendMemberRequest = async (email) => {
    console.log("sendMemberRequest...");
    const targetEmail = await checkUser(email); // find user
    if (!targetEmail) {
      alert(" This user is not registered in the app yet. Perhaps you can share this application with them first.");  
      props.onRequestClose();
      return
    }
    if (targetEmail == user.email) {
      alert("You cannot invite yourself !");
      props.onRequestClose();
      return
    }
    if (targetEmail) {
      console.log("found user, sending request");
      // Cr√©er la notification
      const db = getDatabase();
      let notifRef = ref(db, 'Notifications/');
      const newRef = push(notifRef)
      set(newRef, {
        'message': `${user.email} wants you to join the agenda \"${props.agendaName}\" !`,
        'target': targetEmail,
        'origin': user.email,
        'type': "agendaMemberRequest",
        'agendaName': props.agendaName,
      })
      alert("Notification sent !");
    }
    props.onRequestClose()
    return
  }


  return (
    <>
      <Modal
        animationType="fade"
        transparent={true}
        visible={props.visible}
        onRequestClose={() => {
          props.onRequestClose();
        }}
      >
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <View style={styles.exitButton}>
              <Button
                title="X"
                color={colors.Darker1}
                onPress={() => props.onRequestClose()}
              />
            </View>
            <Text>
              {relations && relations.map((relation, index) => {
                return (
                  <TouchableOpacity style={styles.relationBox} key={index} onPress={() => sendMemberRequest(relation)}>
                    <Text>{relation}</Text>
                  </TouchableOpacity>
                )
              })}</Text>
          </View>
        </View>
      </Modal>
    </>
  );
}