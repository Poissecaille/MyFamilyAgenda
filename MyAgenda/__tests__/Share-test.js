import {ShareComponent} from '../src/components/Share';    
import { expect } from '@jest/globals';
import React from 'react'
import {create} from 'react-test-renderer';
// Note: test renderer must be required after react-native.

const tree= create(
<ShareComponent/>
);

test('test Share component', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });