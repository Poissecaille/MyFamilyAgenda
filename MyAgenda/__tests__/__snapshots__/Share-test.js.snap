// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Share component 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "flex-end",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
        "backgroundColor": "#C8D6F6",
        "borderRadius": 20,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "textAlign": "center",
        "width": 60,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/share2.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "aspectRatio": 1,
          "height": "50%",
          "width": "50%",
        }
      }
    />
  </View>
</View>
`;
